product_state/models/product_template.py

```diff
@@ -1,18 +1,40 @@
 # Copyright 2017 ACSONE SA/NV (<http://acsone.eu>)
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields
+from odoo import api, models, fields
+
+
+class ProductState(models.Model):
+    _name = 'product.state'
+    _description = "Product State"
+    _order = 'name'
+
+    name = fields.Char('State Name', required=True)
+    code = fields.Char('State Code')
+    description = fields.Text(translate=True)
+    product_ids = fields.One2many(
+        'product.template',
+        'product_state_id',
+        string='State Products',
+    )
+    products_count = fields.Integer(
+        string='Number of products',
+        compute='_compute_products_count',
+    )
+
+    @api.multi
+    @api.depends('product_ids')
+    def _compute_products_count(self):
+        for state in self:
+            state.products_count = len(state.product_ids)
```

Comment:
You maybe need to check performance, so instead use a read_group before loop.

Reply:
Please, could you help me more with this?

Reply:
write this
<pre>
data = self.env["product.template"].read_group([
    ('product_state_id', 'in', self.ids),
], ['product_state_id'], ['product_state_id'])
mapped_data  = {
    record['product_state_id'][0]: record['product_state_id_count']
    for record in data
}
for state in self:
    state.products_count = mapped_data.get(state.id, 0)
</pre>